### PR TITLE
New Reconciler Infra

### PR DESCRIPTION
**Pull request by sebmarkbage**
This is an outline for the new reconciler infrastructure.
    
I created a noop renderer to have something to get started from.
    
I split the reconciler folder into old and new, as well as shared. I put shouldUpdateReactComponent in shared as an example of a utility that can easily be shared between both. I plan on breaking out more utilities like these.

Builds on top of #6682

### CODE REVIEW

**Comment by facebook-github-bot-3** [comment]
@sebmarkbage updated the pull request.

**Comment by jayphelps** [comment]
#6170 doesn't really explain these architecture changes. e.g. high vs. low priority renders and incremental and the how and why. Anywhere I can see the discussions on this or have they thus far been informal and undecided and this PR is just to flesh them out? 👯

**Comment by gaearon** [comment]
>have they thus far been informal and undecided and this PR is just to flesh them out?

Yes, I think this is just @sebmarkbage trying to create a useful proof of concept in a way that makes sense to him. The goals are simple: change React from deep recursive rendering on every change to some kind of scheduling.

The way I understand this PR is the following. For scheduling to work, we need some kind of priority system, or otherwise there is no way for React to know which updates should be processed first. “Low” and “high” is the simplest possible way to describe priorities. We don’t know if more kinds of priorities would be helpful, so starting with two makes sense to me. How they are used is an open question, but it’s easy to imagine some options. For example, we might want to schedule high priority updates for something is interacting with (e.g. pressing a Like button) because UI should be responsive to feedback. On the other hand, rendering loading data might be a lower priority (i.e. it’s better to render a new story in the feed with a small delay than cause an animation happening at the same time to drop frames).

If there is any specific algorithm in mind, it’s in @sebmarkbage’s head, but I think he’s just fleshing out a system capable of rendering existing React components with priorities. There’s really not much to it aside from what I wrote above. The existing reconciler code is deeply OO and relies on recursive dynamic dispatches (e.g. `mountComponent()` of any composite component [causes `mountComponent()` of whatever it renders to](https://github.com/facebook/react/blob/b6a60781677bcdace8f41457d50412096f98d74a/src/renderers/shared/reconciler/ReactCompositeComponent.js#L474)) so it’s hard to shove priorities into the existing system. This is why @sebmarkbage is starting the reconciler from scratch in this PR. It’s not clear how long it will take to get to the feature parity, but we’ll see.

**Comment by gaearon** [comment]
Also, for the record, I don’t understand what `Fiber`s are either yet 😄 . The `EffectHandler` and `EffectTag` reminds me of [this post by @sebmarkbage about algebraic effects](https://esdiscuss.org/topic/one-shot-delimited-continuations-with-effect-handlers). The gist of “effects” in [Eff](http://www.eff-lang.org) is that functions can throw “effects” and not just errors, and their callers can “catch” these effects, do something, and then yield back to the child function.

This sounds a lot like what React components want to do. Today, we can say `setState()` is like an effect because it schedules an update to the system. But we want React components to be able to do more and to provide more ways for children and parent to communicate. This is crucial for supporting layout in React: you want a child to “yield” its size to the parent while rendering so that the parent can use this size to set the child’s position. Some notes on a system like this [have been posted in react-future two years ago](https://github.com/reactjs/react-future/blob/ba5a275d08451756137559714d79f15055b9a14e/04%20-%20Layout/02%20-%20Layout%20Components.js) but this wasn’t being worked on. However, two months ago @sebmarkbage, inspired by effects in Eff, made another prototype of a system that would support layout, [also in react-future](https://github.com/reactjs/react-future/blob/ba5a275d08451756137559714d79f15055b9a14e/04%20-%20Layout/prototype/index.js).

Since we don’t actually have algebraic effects in JavaScript, my guess would be that `Fiber`s are some sort of internal abstraction for this pattern of “throwing” effects and having callers handle them. But this may change any day as @sebmarkbage’s still trying different approaches to see what is minimally needed to implement a scheduling reconciler while ensuring future use cases like layout in React are supported better than in the current model.

**Comment by sebmarkbage** [comment]
Fibers as a concept is essentially https://en.m.wikipedia.org/wiki/Fiber_(computer_science) reimplemented in user space.

But yea, I don't really know what I'm doing yet. I have a rough idea.

**Comment by facebook-github-bot-7** [comment]
@sebmarkbage updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@sebmarkbage updated the pull request.

File: src/renderers/noop/__tests__/ReactNoop-test.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    function Bar() {
+      return <div>Hello World</div>;
+    }
+
+    function Foo() {
+      return <Bar isBar={true} />;
+    }
+
+    ReactNoop.render(<Foo />);
+    console.log('Nothing done');
```
Comment (spicyj) [inline src/renderers/noop/__tests__/ReactNoop-test.js]: no (unmocked) logs in tests please

File: src/renderers/shared/fiber/ReactFiberFunctionalComponent.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  console.log('perform work on:', fn.name);
+  var nextElement = fn(props);
+
+  if (typeof nextElement.type === 'function') {
+    return exports.createFiber(nextElement);
+  }
+  return null;
+};
+
+exports.createFiber = function(element : ReactElement) {
```
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberFunctionalComponent.js]: Can you pick different names for createFiber and createFiber here? It's confusing.

**Comment by spicyj** [comment]
Meh, okay.

**Comment by facebook-github-bot-4** [comment]
@sebmarkbage updated the pull request.

**Comment by zpao** [comment]
This (and to a lesser extent the followups) are going to make it tricky to continue cherry-picking into the 15 release due to file moves. Is there any risk in cherry-picking this (and the other fiber changes) as well? It doesn't look like anything in current code has really changed. They aren't part of public modules so seems like it should be pretty safe. We could even exclude the whole `fiber/` dir from the package if we want.

**Comment by zpao** [comment]
After discussing offline - going to take this and some other PRs in 15. While they are currently built in the same pipeline, they aren't part of our exposed API and thus are exempt from semver rules.

**Comment by gaearon** [comment]
>It’s not clear how long it will take to get to the feature parity, but we’ll see.

I guess we know now. 😄

**Pull request merged**